src/components/tooltip/GCMarker.js

```diff
@@ -423,80 +430,104 @@ function _treeInsert(
-    node = { branches: new Map() };
+    const value = { name: '(temp)', time: 0 };
+    node = { branches: new Map(), leaf: false, value };
 
     tree.set(component, node);
   }
 
   if (path.length > 0) {
     // There are more path components.  This node should be a branch if it
     // isn't one already.
-    if (node.value) {
-      // We delete the value to change this node from a leaf to a branch.
-      delete node.value;
-    }
+    node.leaf = false;
     _treeInsert(node.branches, path, phase);
   } else {
     // Make the new node leaf node.
-    if (node.value) {
+    if (node.value.name !== '(temp)') {
       console.error(
-        'Duplicate phases in _treeInsert in MarkerTooltipContents.js'
+        `Duplicate phase ${phase.name} in _treeInsert in MarkerTooltipContents.js`
       );
       return;
     }
     node.value = phase;
+    node.leaf = true;
   }
 }
 
-function _treeGetLeaves(
-  tree: Map<string, PhaseTreeNode>
-): Array<PhaseTimeTuple> {
-  const leaves = [];
+function _treeMapNodes<T>(
+  tree: Map<string, PhaseTreeNode>,
+  f: (node: PhaseTreeNode) => T | null,
+  results: Array<T> = []
+): Array<T> {
   for (const node of tree.values()) {
-    if (node.value) {
-      leaves.push(node.value);
-    } else {
-      leaves.push(..._treeGetLeaves(node.branches));
+    const result = f(node);
+    if (result !== null) {
+      results.push(result);
+    }
+    if (!node.leaf) {
+      _treeMapNodes(node.branches, f, results);
     }
   }
-  return leaves;
+  return results;
 }
 
 function _filterInterestingPhaseTimes(
   rawPhases: PhaseTimes<Microseconds>,
   numSelect: number
 ): Array<PhaseTimeTuple> {
-  let phaseTimes = _makePhaseTimesArray(rawPhases);
+  const phaseTimes = _makePhaseTimesArray(rawPhases);
 
   /*
-   * Select only the leaf phases.
+   * Build the tree.
    */
   const tree = new Map();
   for (const phase of phaseTimes) {
     const components = phase.name.split('.');
     _treeInsert(tree, components, phase);
   }
-  phaseTimes = _treeGetLeaves(tree);
 
   /*
-   * Of those N leaf phases, select the M most interesting phases by
-   * determining the threshold we want to stop including phases at.
-   *
-   * Calculate the threshold by sorting the list of times in asscending
-   * order, then slicing off all the low items and looking at the first
-   * item.
+   * For every non-leaf phase, add in a (self-time) leaf with the time not
+   * accounted for by its children.
    */
-  const sortedPhaseTimes = phaseTimes
-    .map(pt => pt.time)
-    // Descending order
-    .sort((a, b) => b - a);
-  const threshold = sortedPhaseTimes[numSelect];
+  _treeMapNodes(
+    tree,
+    ({ leaf, branches, value: { time: totalTime, name } }) => {
+      if (!leaf) {
+        let remaining = totalTime;
+        for (const { value } of branches.values()) {
+          remaining -= value.time;
+        }
+        const selfTimeName = `${name} (self-time)`;
```

Comment:
What do you think of extracting this ` (self-time)` string to a constant, so that it can also be reused below, and we don't run into problems in the future?